core/modules/link/link.module

```diff
@@ -335,6 +349,11 @@ function link_field_widget_form(&$form, &$form_state, $field, $instance, $langco
     '#type' => $instance['widget']['type'],
     '#default_value' => isset($items[$delta]) ? $items[$delta] : '',
   );
+  if ($instance['settings']['type'] != LINK_BOTH) {
+    $element['#description'] .= ' ' . t('Only %type URLs are allowed.', array(
```

Comment:
We should probably separate this out into two separate strings so it can be fully translated. Right now "internal" and "external" would be used regardless of translation. I would type out the entire string to provide context to translators:

```
$instance['settings']['type'] == LINK_INTERNAL ? t('Only internal paths are allowed.') : t('Only full URLs are allowed.);
```

Though I might lean towards not putting any automatic help text at all, like @indigoxela suggested.